File: core/src/main/kotlin/br/com/dillmann/nginxignition/core/accesslist/AccessListService.kt
package br.com.dillmann.nginxignition.core.accesslist

import br.com.dillmann.nginxignition.core.accesslist.command.DeleteAccessListByIdCommand
import br.com.dillmann.nginxignition.core.accesslist.command.GetAccessListByIdCommand
import br.com.dillmann.nginxignition.core.accesslist.command.ListAccessListCommand
import br.com.dillmann.nginxignition.core.accesslist.command.SaveAccessListByCommand
import br.com.dillmann.nginxignition.core.common.pagination.Page
import br.com.dillmann.nginxignition.core.host.HostRepository
import java.util.*

internal class AccessListService(
    private val accessListRepository: AccessListRepository,
    private val hostRepository: HostRepository,
    private val validator: AccessListValidator,
): DeleteAccessListByIdCommand, GetAccessListByIdCommand, ListAccessListCommand, SaveAccessListByCommand {
    override suspend fun deleteById(id: UUID): DeleteAccessListByIdCommand.Output {
        if (hostRepository.existsByAccessListId(id))
            return DeleteAccessListByIdCommand.Output(
                false,
                "Access list is being used by at least one host",
            )

        accessListRepository.deleteById(id)
        return DeleteAccessListByIdCommand.Output(true, "Access list deleted successfully")
    }

    override suspend fun getById(id: UUID): AccessList? =
        accessListRepository.findById(id)

    override suspend fun getPage(pageNumber: Int, pageSize: Int, searchTerms: String?): Page<AccessList> =
        accessListRepository.findPage(pageNumber, pageSize, searchTerms)

    override suspend fun save(accessList: AccessList) {
        validator.validate(accessList)
        accessListRepository.save(accessList)
    }
}

File: core/access_list/access_list_service.go
package access_list

import (
	"dillmann.com.br/nginx-ignition/core/common/pagination"
	"dillmann.com.br/nginx-ignition/core/host"
	"errors"
	"github.com/google/uuid"
)

type accessListService struct {
	accessListRepository *AccessListRepository
	hostRepository       *host.HostRepository
}

func (service *accessListService) save(accessList *AccessList) error {
	err := newValidator().validate(accessList)
	if err != nil {
		return err
	}

	return (*service.accessListRepository).Save(accessList)
}

func (service *accessListService) deleteById(id uuid.UUID) error {
	inUse, err := (*service.hostRepository).ExistsByAccessListId(id)
	if err != nil {
		return err
	}

	if inUse {
		return errors.New("access List is in use by one or more hosts")
	}

	return (*service.accessListRepository).DeleteById(id)
}

func (service *accessListService) findById(id uuid.UUID) (*AccessList, error) {
	return (*service.accessListRepository).FindById(id)
}

func (service *accessListService) findAll() (*[]AccessList, error) {
	return (*service.accessListRepository).FindAll()
}

func (service *accessListService) list(
	pageNumber int64,
	pageSize int64,
	searchTerms string,
) (*pagination.Page[AccessList], error) {
	return (*service.accessListRepository).FindPage(pageNumber, pageSize, searchTerms)
}
